### ISSUE TITLE
Official Docker Image

### ISSUE THREAD

**Issue by pozgo** [mkdocs/mkdocs#2136]
I'm the author of one of the popular docker images with mkdocs and was wondering if we maybe should join forces on that front and I could help/lead an official docker image for mkdocs if you feel this is something you would like to offer to the community. 

I have lead many opensource docker images projects and can offer my assistance with this. Please see the links to my repos and docker hub. 

mkdocs:
https://hub.docker.com/r/polinux/mkdocs
https://github.com/pozgo/docker-mkdocs

GH
https://github.com/pozgo
https://github.com/pozgo?tab=repositories&q=docker&type=&language=
https://github.com/million12

**Comment by rockandska** [comment]
Any news about this ?

**Comment by oprypin** [comment]
I don't understand the use case for this.

Say you do have an image with MkDocs. Then you'll need to install a theme and some plugins anyway. So why not just install mkdocs as well while you're at it?

(Certainly I wouldn't want requests for a curated selection of plugins to pop up.)

*And* you get to pick whatever base image you like. (what if for some plugin you need a particular Python version)

---

So with that in mind, this would conflict with my personal opinion: that is, how could I in good conscience endorse a Docker image, if my actual recommendation is to install MkDocs without Docker?

---

And beside my main points, I also don't see anyone from MkDocs wanting to maintain this.

So pretty much it's just not happening.

**Comment by pbendersky** [comment]
Just my 2 cents on a use case... I came across this issue, since I want to generate the docs on every commit and host them on GitLab Pages. The CI process runs with a Docker image, so my first instinct was to look for an official one.

**Comment by colbygk** [comment]
> 

Providing docker resources can range from an official image to a Dockerfile checked into a repo. Having a Dockerfile that maps in directories into the image can make a flexible container that allows configuration of both themes and plugins.

Mean while, if someone needs a different version of python, they can use the Dockerfile and sub in their own base layers. Having an official image also helps avoid mkdocs ending up as a vector of malware. This helps isolate concerns without forcing one to install mkdocs on a system that may need to co-exist with other versions of python without resorting to virtualenv and separate pip libraries (yet more vectors of malware).

**Comment by oprypin** [comment]
Thanks for your arguments. My point still stands, though. "mkdocs image + `pip install mkdocs-some-plugin`" to me has only disadvantages compared to "python image + `pip install mkdocs mkdocs-some-plugin`.

**Comment by SchulteMarkus** [comment]
+1
I think for the named arguments an official docker image (and maybe a GitHub action) makes sense. People using mkdocs without any additional plugin will be saved a lot of effort (for example to host their MkDocs doc as GitHub page).

**Comment by oprypin** [comment]
> People using mkdocs without any additional plugin will be saved a lot of effort (for example to host their MkDocs doc as GitHub page).

You're stating it like it's just a fact, but could you tell me exactly in what ways it would be less effort than just putting this exact file into your repository?:
https://github.com/oprypin/mkdocs-literate-nav/blob/master/.github/workflows/deploy-docs.yml

**Comment by SchulteMarkus** [comment]
> You're stating it like it's just a fact, but could you tell me exactly in what ways it would be less effort than just putting this exact file into your repository?: https://github.com/oprypin/mkdocs-literate-nav/blob/master/.github/workflows/deploy-docs.yml

Ok. You got a point (it could probably be shorter, but hey, never mind).
Though I still think an official docker image would be appreciated by devs being able to install Python but having Docker available.

**Comment by pawamoy** [comment]
But if they have Docker, they can just pull a Python image and pip install MkDocs, no? Maybe they dont have access to pypi.org for security reasons? Then they shouldnt have access to docker hub either, for the same reason ^^ I fail to find a scenario where an MkDocs image is necessary 😕

**Comment by SchulteMarkus** [comment]
> But if they have Docker, they can just pull a Python image and pip install MkDocs, no? Maybe they dont have access to pypi.org for security reasons? Then they shouldnt have access to docker hub either, for the same reason ^^ I fail to find a scenario where an official MkDocs image is necessary confused

Well.. ok
Docker would "only" be convenient.
But if convenience is to be avoided, someone could argue, why upload MkDocs to pypi? Packaged zip could be downloaded from GitHub release assets easily.
And I have to say: Having a Dockerfile as part of this project and hosting of the Docker image using GitHub seems not to be a big deal. I can provide a PR, shall I?

**Comment by oprypin** [comment]
Please no pull request at this time. There are no arguments for it and nobody willing to take up maintenance. Even picking the base Python image is a big deal. I don't want issues like "oh, why did you already update the Python version, my plugin breaks with it"

**Comment by colbygk** [comment]
I am still trying to understand this:

How does https://github.com/oprypin/mkdocs-literate-nav/blob/master/.github/workflows/deploy-docs.yml affect this discussion?

**Comment by colbygk** [comment]
> Say you do have an image with MkDocs. Then you'll need to install a theme and some plugins anyway. So why not just install mkdocs as well while you're at it?

I don't follow this argument. Having an official image would come with a way of providing access to plugins and themes, it doesn't stop this in any way while adding capability/ease of use. Docker is a larger universe than python+mkdocs itself. Docker is on the level of abstraction that makes it runable anywhere without having to worry if a particular environment supports python or not. It is literally the most portable way of doing this.

> And you get to pick whatever base image you like. (what if for some plugin you need a particular Python version)

This is also true of a docker file. Nothing stops this from being possible and and image would provide a known correct combination of python+mkdocs. Anyone who wants to make something different can use the underlying Dockerfile to configure an image with a different python version. This increases the ability to provide a stable and supportable environment runtime.

**Comment by oprypin** [comment]
OK. The official Docker image for MkDocs is 
https://hub.docker.com/_/python

If MkDocs doesn't actually work on that at any point, the bug will be treated at the highest priority.

I have already said all my arguments. Mainly that forcing a choice of Python version and OS is a much bigger disadvantage that saving you typing `pip install mkdocs`. Now the conversation is just repeating itself.

> without having to worry if a particular environment supports python or not. It is literally the most portable way of doing this.

This solves that too 👌👌👌

**Comment by colbygk** [comment]
> forcing a choice of Python version and OS is a much bigger disadvantage that saving you typing pip install mkdocs

This is a profound misunderstanding of what Docker is and how developers and system administrators use it to their advantage in terms of speed and reliability of developing and reproducibly standing up services.

**Comment by pozgo** [comment]
Ok Since I'm the author of the image let me add something. 

If there was no need for an automated docker image with a possibility to fetch changes of documentation from git my image would have no downloads. Yet there are already half a million downloads. 

Another thing is that there are already 27 forks so it seems people actually using them. And I was using it for a very long time to a degree where we had it in the Kubernetes cluster with internal docs.

And I'm a maintainer of that image. If we migrated it to the official MKDocs repo I can be of assistance to keep an eye on it. It's not that it's being updated 20 times a day, is it? 

Either way, feel free to let me know if you want to make this a part of the official MKDocs. 

Peace

**Comment by oprypin** [comment]
Thanks. But seems it's much easier to just keep using https://github.com/pozgo/docker-mkdocs - you can update it yourself without the overhead of my uninformed code reviews.

**Comment by tomchristie** [comment]
Thanks @pozgo and @christianechevarria.

**Comment by CieNTi** [comment]
First of all, thanks to all MkDocs team for this amazing tool, and big thanks to @pozgo for the containerized version of it, what a great job.

I'm a long-time user, using my own docker image since the very first day, and I just arrived here one more time just to see if an official image is finally released.

It is a pity the opposition to provide an _official_ docker image using what, under my opinion, are so weak pillars, instead of looking the great job and knowledge from @pozgo.

In this thread it is assumed the users will have a working python environment, but that does not need to be true at all, and only forces people to install something that maybe is just needed to be able to run MkDocs, while docker covers orders of magnitude more scenarios in a cleaner way.

Taking into consideration that MkDocs is more an application that a package itself, by providing a docker image you are providing not only a package or an application, but a _working environment_ where the developers ensures the right working conditions, so everything is already packed (containerized) and ready to run.

The recommendation of "just do `pip install mkdocs`" is a bad recommendation in terms of reproducibility, because there is no version pinning, and yes, it is possible to pin by command and by requirements, but is way more easier, cleaner and safer to just do `docker run mkdocs:latest` to match the same recommendation, but also `docker run mkdocs:1.2.3` to match _your_ already tested version, without previously install python, pip, mkdocs ...

"Some plugins may fail because this or that OS" is not a MkDocs problem, and it is already answered in the official MkDocs readme: _Please note that we may only provide support for problems/questions regarding core features of MkDocs. Any questions or bug reports about features of third-party themes, plugins, extensions or similar should be made to their respective projects._

Apart of that, from a developer point of view, by using containers instead of system installation it is ensured any person will have the very same scenario (developer and end-user), avoiding the _it worked in my PC_ problem.

Please, review this case and take into consideration to provide an official image, more when the work is already done by @pozgo. Not everyone must know about his work, because people come to the official place for official things.

**Comment by pawamoy** [comment]
If I had the power to do so, I'd be glad to start endorsing third-party projects, and even moving them into the org to make them official, while letting original authors maintain them.

We actually started such an initiative a couple months ago. Lots of people in the ecosystem where excited and interested in participating. But almost nothing happened since then.

**Comment by mrmanc** [comment]
FWIW there’s an [official Material Docker image](https://hub.docker.com/r/squidfunk/mkdocs-material/) which might help users. It’s highly popular, with 5M+ pulls.

**Comment by pozgo** [comment]
Well. Literally no one contacted me about this :D As I stated years back that I'm more than happy to be main contributor and maintainer of official docker container for `mkdocs` I still stand my ground. The image I'm providing is partially maintain by me and some people around. It would be great to create a dedicated  team to build that image even better. I never pushed this to be fully CI/CD managed and tested since lack of resources and time. 

What we have ATM is working version of docker container but need a bit of love not only from myself but `mkdocs` itself. w
I'm an author/co-author of many official images, directly or in-directly (grafana/gitlac-ce ande a few more)

So TLDR

If you want first community project to start ping me and add me to an organisation and we can make this happen. Whats the better way that be an example of started and finished project? :D

**Comment by squidfunk** [comment]
Maintainer of Material for MkDocs here! We've been happily providing the official Material for MkDocs Docker image since version 1.0.5 (7 years ago) with huge success, given it already has 7.3M downloads! We believe that albeit MkDocs is written in Python, users should not be concerned with getting up and running with Python, which can be tedious. I've been a beginner myself 8 years ago and know how frustrating it can be 😅

@tomchristie would you consider giving @pozgo sufficient permissions to work on the Docker image and advertise under the MkDocs organization? IMHO, a Docker image would be of great value, expanding the user base of MkDocs ☺️ Maybe we can create a separate project in the organization and make @pozgo a maintainer? Also, removing the warning is something that has been wished for for a long time, see https://github.com/mkdocs/mkdocs/issues/2108

**Comment by pawamoy** [comment]
> Well. Literally no one contacted me about this :D

Sorry, I only tagged a handful of plugin/extension authors back then 😅 

> It would be great to create a dedicated team to build that image even better. I never pushed this to be fully CI/CD managed and tested since lack of resources and time.

> What we have ATM is working version of docker container but need a bit of love not only from myself but mkdocs itself.

Please note that even if MkDocs endorses the Docker image you maintain by bringing it into the org, it doesn't mean you'll suddenly have more resources at your disposal. You'll still probably only get help from the community. The increased visibility brought by the endorsement could mean more involvement from the community though 👍 Well, I'm mainly speaking for myself here, please anyone correct me if I'm wrong.

Anyway, thanks for your continued work on the Docker image and for keeping your offer on the table 🙂

**Comment by tomchristie** [comment]
> would you consider giving @pozgo sufficient permissions to work on the Docker image

What permissions are needed? My assumptions would be that the intent is to be able to publish to `https://hub.docker.com/r/mkdocs/mkdocs/`, and that some repo permissions are required in order to do so?

What does maintaining the docker image involve?

**Comment by pozgo** [comment]
I guess that depends how we want to approach the problem. We can either use the main repo of `mkdocs` to maintain it and follow the same cycle of releases, or we can have a separate repo that builds the image fetching data from the main repo and publishes it to docker hub. I think that would require a bit of discussion on how we want to do this and quite possibly come up with even other solution. 

There is few things I want to improve in my image too just need to plan when I want to do this. And having a community support with the project would help a lot as we have very talented people willing to chip in. 

I prefer clean and easy ways of building and publishing without too much of complexity - it's much easier to maintain. 

Other question is what features we want to keep within the image itself. 
My image have auto download from a github repo which is extra feature. Possibly not needed in official. I think offcial should be pretty much vanilla and if we decide we want to have extra features we could add them but only if the community really needs them. 

So from my side: I'm going to revisit the image and see what needs fixing and try tu update it to the latest version of `mkdocs` and in the mean time we can set something in motion regarding the official image. (Unless we agree to move my image to official as is.)
